src/config.py

```diff
@@ -0,0 +1,168 @@
+'''
+This config file is being used as a singleton. Because python creates a single instance
+of modules when they're imported, we rely on the Snakefile instantiating the module.
+In particular, when the Snakefile calls init_config, it will reassign config
+to take the value of the actual config provided by Snakemake. After that point, any
+module that imports this module can access a config option by checking the object's
+value. For example
+
+import spras.config as config
+container_framework = config.config.framework
+
+will grab the top level registry configuration option as it appears in the config file
+'''
+import itertools as it
+import numpy as np
+import yaml
+import copy as copy
+import os
+from src.util import hash_params_sha1_base32
+
+# The default length of the truncated hash used to identify parameter combinations
+DEFAULT_HASH_LENGTH = 7
+DEFAULT_CONTAINER_PREFIX = "docker.io/reedcompbio"
+
+config = None
+
+# This will get called in the Snakefile, instantiating the singleton with the raw config
+def init_global(config_dict):
+    global config
+    config = Config(config_dict)
+
+def init_from_file(filepath):
+    global config
+
+    # Handle opening the file and parsing the yaml
+    filepath = os.path.abspath(filepath)
+    try:
+        with open(filepath, 'r') as yaml_file:
+            config_dict = yaml.safe_load(yaml_file)
+    except FileNotFoundError:
+        print(f"Error: The specified config '{filepath}' could not be found.")
+        return False
+    except yaml.YAMLError as e:
+        print(f"Error: Failed to parse config '{filepath}': {e}")
+        return False
+    
+    # And finally, initialize
+    config = Config(config_dict)
+
+
+class Config:
+    def __init__(self, raw_config):
+        # Since process_config winds up modifying the raw_config passed to it as a side effect,
+        # we'll make a deep copy here to guarantee we don't break anything. This preserves the 
+        # config as it's given to the Snakefile by Snakemake
+        _raw_config = copy.deepcopy(raw_config)
+        self.process_config(_raw_config)
+
+    def process_config(self, raw_config):
```

Comment:
The essence of `process_config` hasn't changed too much. The old `process_config` from `util.py` created a variety of variables to be returned, namely `config, datasets, out_dir, algorithm_params, algorithm_directed, pca_params, hac_params`. Rather than return anything, the new `process_config` instead makes these accessible as member variables of the underlying config class. When the global config object is instantiated by Snakemake as it runs through the Snakefile, these member variables should all be created.

One other change is that the old `process_config` wound up modifying the config dict that was passed to it by the Snakefile. Those modifications were used later in the Snakefile to handle setting up analysis parameters (see [here](https://github.com/Reed-CompBio/spras/blob/e67167fcc880973421847a94bbd871317951f683/Snakefile#L60C5-L60C19)). I decided to handle the `TODO` on that line and also parse that info as a part of the config object. This negated the need to modify the underlying config dict as created by Snakemake, so I also added a deepcopy of the config that gets passed to `process_config`, guaranteeing that the `config` variable available in the Snakefile remains unchanged.